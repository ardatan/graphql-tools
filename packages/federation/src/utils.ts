import { GraphQLSchema } from 'graphql';
import { FilterRootFields, FilterTypes } from '@graphql-tools/wrap';

export function getArgsFromKeysForFederation(representations: readonly any[]) {
  return { representations };
}

export function getKeyForFederation(root: any) {
  return root;
}

export function filterInternalFieldsAndTypes(finalSchema: GraphQLSchema) {
  const removeEntitiesField = new FilterRootFields(
    (operation, fieldName) =>
      !(operation === 'Query' && (fieldName === '_entities' || fieldName === '_sdl')),
  );
  const removeEntityAndAny = new FilterTypes(
    type =>
      type.name !== '_Entity' &&
      type.name !== '_Any' &&
      type.name !== '_FieldSet' &&
      type.name !== '_Service' &&
      !type.name.startsWith('link__'),
  );
  const fakeSubschemaConfig = {
    schema: finalSchema,
    transforms: [removeEntitiesField, removeEntityAndAny],
  };
  for (const transform of fakeSubschemaConfig.transforms) {
    finalSchema = transform.transformSchema(finalSchema, fakeSubschemaConfig);
  }
  return finalSchema;
}
