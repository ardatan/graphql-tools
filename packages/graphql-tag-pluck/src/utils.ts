// Will use the shortest indention as an axis
export const freeText = (text: string | string[], skipIndentation = false) => {
  if (text instanceof Array) {
    text = text.join('');
  }

  // This will allow inline text generation with external functions, same as ctrl+shift+c
  // As long as we surround the inline text with ==>text<==
  text = text.replace(/( *)==>((?:.|\n)*?)<==/g, (_match, baseIndent, content: string) => {
    return content
      .split('\n')
      .map(line => `${baseIndent}${line}`)
      .join('\n');
  });

  if (skipIndentation) {
    if (text.trim() === '') {
      return '';
    }
    return text;
  }

  const lines = text.split('\n');

  const minIndent = lines
    .filter(line => line.trim())
    .reduce<number>((minIndent, line) => {
      const currIndent = line.match(/^ */)?.[0].length;
      if (currIndent == null) {
        return minIndent;
      }

      return currIndent < minIndent ? currIndent : minIndent;
    }, Infinity);

  return lines
    .map(line => line.slice(minIndent))
    .join('\n')
    .trim()
    .replace(/\n +\n/g, '\n\n');
};

// foo_barBaz -> ['foo', 'bar', 'Baz']
export const splitWords = (str: string) => {
  return str.replace(/[A-Z]/, ' $&').split(/[^a-zA-Z0-9]+/);
};

// upper -> Upper
export const toUpperFirst = (str: string) => {
  return str.slice(0, 1).toUpperCase() + str.slice(1).toLowerCase();
};

// foo-bar-baz -> fooBarBaz
export const toCamelCase = (str: string) => {
  const words = splitWords(str);
  const first = words.shift()?.toLowerCase() ?? '';
  const rest = words.map(toUpperFirst);

  return [first, ...rest].join('');
};
