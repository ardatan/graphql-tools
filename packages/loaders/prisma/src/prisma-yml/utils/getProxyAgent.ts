'use strict';

import HttpsProxyAgent from 'https-proxy-agent';
import HttpProxyAgent from 'http-proxy-agent';
import { Agent as HttpsAgent } from 'https';
import { Agent as HttpAgent } from 'http';

// code from https://raw.githubusercontent.com/request/request/5ba8eb44da7cd639ca21070ea9be20d611b85f66/lib/getProxyFromURI.js

function formatHostname(hostname: any) {
  // canonicalize the hostname, so that 'oogle.com' won't match 'google.com'
  return hostname.replace(/^\.*/, '.').toLowerCase();
}

function parseNoProxyZone(zone: any) {
  zone = zone.trim().toLowerCase();

  const zoneParts = zone.split(':', 2);
  const zoneHost = formatHostname(zoneParts[0]);
  const zonePort = zoneParts[1];
  const hasPort = zone.indexOf(':') > -1;

  return { hostname: zoneHost, port: zonePort, hasPort };
}

function uriInNoProxy(uri: any, noProxy: any) {
  const port = uri.port || (uri.protocol === 'https:' ? '443' : '80');
  const hostname = formatHostname(uri.hostname);
  const noProxyList = noProxy.split(',');

  // iterate through the noProxyList until it finds a match.
  return noProxyList.map(parseNoProxyZone).some(function (noProxyZone: any) {
    const isMatchedAt = hostname.indexOf(noProxyZone.hostname);
    const hostnameMatched = isMatchedAt > -1 && isMatchedAt === hostname.length - noProxyZone.hostname.length;

    if (noProxyZone.hasPort) {
      return port === noProxyZone.port && hostnameMatched;
    }

    return hostnameMatched;
  });
}

function getProxyFromURI(uri: any) {
  // Decide the proper request proxy to use based on the request URI object and the
  // environmental variables (NO_PROXY, HTTP_PROXY, etc.)
  // respect NO_PROXY environment variables (see: http://lynx.isc.org/current/breakout/lynx_help/keystrokes/environments.html)

  const noProxy = process.env['NO_PROXY'] || process.env['no_proxy'] || '';

  // if the noProxy is a wildcard then return null

  if (noProxy === '*') {
    return null;
  }

  // if the noProxy is not empty and the uri is found return null

  if (noProxy !== '' && uriInNoProxy(uri, noProxy)) {
    return null;
  }

  // Check for HTTP or HTTPS Proxy in environment Else default to null

  if (uri.protocol === 'http:') {
    return process.env['HTTP_PROXY'] || process.env['http_proxy'] || null;
  }

  if (uri.protocol === 'https:') {
    return (
      process.env['HTTPS_PROXY'] ||
      process.env['https_proxy'] ||
      process.env['HTTP_PROXY'] ||
      process.env['http_proxy'] ||
      null
    );
  }

  // if none of that works, return null
  // (What uri protocol are you using then?)

  return null;
}

export function getProxyAgent(url: string): HttpAgent | HttpsAgent | undefined {
  const uri = new URL(url);
  const proxy = getProxyFromURI(uri);
  if (!proxy) {
    return undefined;
  }

  const proxyUri = new URL(proxy);

  if (proxyUri.protocol === 'http:') {
    // eslint-disable-next-line
    // @ts-ignore
    return new HttpProxyAgent(proxy) as HttpAgent;
  }

  if (proxyUri.protocol === 'https:') {
    // eslint-disable-next-line
    // @ts-ignore
    return new HttpsProxyAgent(proxy) as HttpsAgent;
  }

  return undefined;
}
