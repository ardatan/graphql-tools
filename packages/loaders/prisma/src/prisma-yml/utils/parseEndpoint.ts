import { clusterEndpointMapReverse } from '../constants.js';
import { URL } from 'url';

function getClusterName(origin: any): string {
  if (clusterEndpointMapReverse[origin]) {
    return clusterEndpointMapReverse[origin];
  }

  if (origin.endsWith('prisma.sh')) {
    return origin.split('_')[0].replace(/https?:\/\//, '');
  }

  if (isLocal(origin)) {
    return 'local';
  }

  return 'default';
}

const getWorkspaceFromPrivateOrigin = (origin: string) => {
  const split = origin.split('_');
  if (split.length > 1) {
    return split[1].split('.')[0];
  }

  return null;
};

const isLocal = (origin: any) => origin.includes('localhost') || origin.includes('127.0.0.1');

export interface ParseEndpointResult {
  service: string;
  clusterBaseUrl: string;
  stage: string;
  isPrivate: boolean;
  local: boolean;
  shared: boolean;
  workspaceSlug: string | null;
  clusterName: string;
}

export function parseEndpoint(endpoint: string): ParseEndpointResult {
  /*
    Terminology:
      local - hosted locally using docker and accessed using localhost or prisma or local web proxy like domain.dev
      shared - demo server
      isPrivate - private hosted by Prisma or private and self-hosted, important that in our terminology a local server is not private
  */

  const url = new URL(endpoint);
  const splittedPath = url.pathname.split('/');
  // assuming, that the pathname always starts with a leading /, we always can ignore the first element of the split array
  const service = splittedPath.length > 3 ? splittedPath[2] : splittedPath[1] || 'default';
  const stage = splittedPath.length > 3 ? splittedPath[3] : splittedPath[2] || 'default';

  // This logic might break for self-hosted servers incorrectly yielding a "workspace" simply if the UX has
  // enough "/"es like if https://custom.dev/not-a-workspace/ is the base Prisma URL then for default/default service/stage
  // pair. This function would incorrectly return not-a-workspace as a workspace.
  let workspaceSlug = splittedPath.length > 3 ? splittedPath[1] : null;

  const shared = ['eu1.prisma.sh', 'us1.prisma.sh'].includes(url.host);

  // When using localAliases, do an exact match because of 'prisma' option which is added for local docker networking access
  const localAliases = ['localhost', '127.0.0.1', 'prisma'];
  const isPrivate = !shared && !localAliases.includes(url.hostname);
  const local = !shared && !isPrivate && !workspaceSlug;

  if (isPrivate && !workspaceSlug) {
    workspaceSlug = getWorkspaceFromPrivateOrigin(url.origin);
  }

  return {
    clusterBaseUrl: url.origin,
    service,
    stage,
    local,
    isPrivate,
    shared,
    workspaceSlug,
    clusterName: getClusterName(url.origin),
  };
}
