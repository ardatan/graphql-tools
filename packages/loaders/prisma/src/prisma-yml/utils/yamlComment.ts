import * as yamlParser from 'yaml-ast-parser';

/**
 * Comments out the current entry of a specific key in a yaml document and creates a new value next to it
 * @param key key in yaml document to comment out
 * @param newValue new value to add in the document
 */
export function replaceYamlValue(input: any, key: any, newValue: any) {
  const ast = yamlParser.safeLoad(input);
  const position = getPosition(ast, key);
  const newEntry = `${key}: ${newValue}\n`;
  if (!position) {
    return input + '\n' + newEntry;
  }

  return (
    input.slice(0, position.start) +
    '#' +
    input.slice(position.start, position.end) +
    newEntry +
    input.slice(position.end)
  );
}

function getPosition(ast: any, key: string): { start: number; end: number } | undefined {
  const mapping = ast.mappings.find((m: any) => m.key.value === key);
  if (!mapping) {
    return undefined;
  }
  return {
    start: mapping.startPosition,
    end: mapping.endPosition + 1,
  };
}

function commentOut(input: string, keys: string[]) {
  let output = input;
  for (const key of keys) {
    const ast = yamlParser.safeLoad(output);
    const position = getPosition(ast, key);

    if (position) {
      output = output.slice(0, position.start) + '#' + output.slice(position.start);
    }
  }

  return output;
}

export function migrateToEndpoint(input: any, endpoint: any) {
  const output = commentOut(input, ['service', 'stage', 'cluster']);
  return replaceYamlValue(output, 'endpoint', endpoint);
}
