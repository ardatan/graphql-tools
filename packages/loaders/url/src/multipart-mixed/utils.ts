/* eslint-disable */

export function getBoundaryFromContentType(contentType: string) {
  const idxBoundary = contentType.indexOf('boundary=');
  return `--${
    !!~idxBoundary
      ? // +9 for 'boundary='.length
        contentType
          .substring(idxBoundary + 9)
          .trim()
          .replace(/['"]/g, '')
      : '-'
  }`;
}

export const separator = '\r\n\r\n';

export const decoder = new TextDecoder();
