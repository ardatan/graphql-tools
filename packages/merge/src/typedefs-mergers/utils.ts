import { Kind, ListTypeNode, NamedTypeNode, NonNullTypeNode, Source, TypeNode } from 'graphql';

export function isStringTypes(types: any): types is string {
  return typeof types === 'string';
}

export function isSourceTypes(types: any): types is Source {
  return types instanceof Source;
}

export function extractType(type: TypeNode): NamedTypeNode {
  let visitedType = type;
  while (visitedType.kind === Kind.LIST_TYPE || visitedType.kind === 'NonNullType') {
    visitedType = visitedType.type;
  }
  return visitedType;
}

export function isWrappingTypeNode(type: TypeNode): type is ListTypeNode | NonNullTypeNode {
  return type.kind !== Kind.NAMED_TYPE;
}

export function isListTypeNode(type: TypeNode): type is ListTypeNode {
  return type.kind === Kind.LIST_TYPE;
}

export function isNonNullTypeNode(type: TypeNode): type is NonNullTypeNode {
  return type.kind === Kind.NON_NULL_TYPE;
}

export function printTypeNode(type: TypeNode): string {
  if (isListTypeNode(type)) {
    return `[${printTypeNode(type.type)}]`;
  }

  if (isNonNullTypeNode(type)) {
    return `${printTypeNode(type.type)}!`;
  }

  return type.name.value;
}

export enum CompareVal {
  A_SMALLER_THAN_B = -1,
  A_EQUALS_B = 0,
  A_GREATER_THAN_B = 1,
}
export type CompareFn<T> = (a: T | undefined, b: T | undefined) => -1 | 0 | 1;

export function defaultStringComparator(a: string | undefined, b: string | undefined): CompareVal {
  if (a == null && b == null) {
    return CompareVal.A_EQUALS_B;
  }
  if (a == null) {
    return CompareVal.A_SMALLER_THAN_B;
  }
  if (b == null) {
    return CompareVal.A_GREATER_THAN_B;
  }
  if (a < b) return CompareVal.A_SMALLER_THAN_B;
  if (a > b) return CompareVal.A_GREATER_THAN_B;
  return CompareVal.A_EQUALS_B;
}
