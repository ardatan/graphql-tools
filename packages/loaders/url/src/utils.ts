import { memoize1 } from '@graphql-tools/utils';
import { OperationDefinitionNode } from 'graphql';
import type { Readable } from 'stream';

export function isBlob(obj: any): obj is Blob {
  return typeof obj.arrayBuffer === 'function';
}

interface GraphQLUpload {
  filename: string;
  mimetype: string;
  createReadStream: () => Readable;
}

export function isGraphQLUpload(upload: any): upload is GraphQLUpload {
  return typeof upload.createReadStream === 'function';
}

export function isPromiseLike(obj: any): obj is PromiseLike<any> {
  return typeof obj.then === 'function';
}

export const isLiveQueryOperationDefinitionNode = memoize1(function isLiveQueryOperationDefinitionNode(
  node: OperationDefinitionNode
) {
  return node.directives?.some(directive => directive.name.value === 'live');
});

export enum LEGACY_WS {
  CONNECTION_INIT = 'connection_init',
  CONNECTION_ACK = 'connection_ack',
  CONNECTION_ERROR = 'connection_error',
  CONNECTION_KEEP_ALIVE = 'ka',
  START = 'start',
  STOP = 'stop',
  CONNECTION_TERMINATE = 'connection_terminate',
  DATA = 'data',
  ERROR = 'error',
  COMPLETE = 'complete',
}
