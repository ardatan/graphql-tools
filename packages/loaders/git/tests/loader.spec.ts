import { execSync } from 'child_process';

import { GitLoader } from '../src/index.js';
import { runTests } from '../../../testing/utils.js';

describe('GitLoader', () => {
  const loader = new GitLoader();
  const lastCommit = execSync('git rev-parse HEAD', { encoding: 'utf-8' }).replace(/\n/g, '');
  const getPointer = (fileName: string) => {
    return `git:${lastCommit}:packages/loaders/git/tests/test-files/${fileName}`;
  };

  describe('canLoad', () => {
    runTests({
      async: loader.canLoad.bind(loader),
      sync: loader.canLoadSync.bind(loader),
    })(canLoad => {
      it('should return true for a valid pointer', async () => {
        await expect(canLoad(getPointer('some-file.graphql'))).resolves.toBe(true);
      });

      it('should return false if pointer does not begin with "git:"', async () => {
        await expect(canLoad(getPointer('some-file.graphql').substring(4))).resolves.toBe(false);
      });

      it('should return false if pointer is not a string', async () => {
        await expect(canLoad(42 as any)).resolves.toBe(false);
      });
    });
  });

  describe('load', () => {
    runTests({
      async: loader.load.bind(loader),
      sync: loader.loadSync.bind(loader),
    })(load => {
      it('should load document from a .graphql file', async () => {
        const [result] = await load(getPointer('type-defs.graphql'), {});
        expect(result.document).toBeDefined();
      });

      it('should load introspection data from a .json file', async () => {
        const [result] = await load(getPointer('introspection.json'), {});
        expect(result.schema).toBeDefined();
      });

      it('should load type definitions from a .json file', async () => {
        const [result] = await load(getPointer('type-defs.json'), {});
        expect(result.document).toBeDefined();
      });

      it('should load type definitions from a pluckable file', async () => {
        const [result] = await load(getPointer('pluckable.ts'), {});
        expect(result.document).toMatchSnapshot();
      });

      it('should throw when the file does not exist', async () => {
        await expect(load(getPointer('wrong-filename.graphql'), {})).rejects.toThrowError(
          'Unable to load file from git'
        );
      });

      it('should simply ignore a non git path', async () => {
        const result = await load('./pluckable.ts', {});
        expect(result).toEqual([]);
      });
    });
  });
});
