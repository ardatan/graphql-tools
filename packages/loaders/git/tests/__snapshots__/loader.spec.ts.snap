// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitLoader load async should load type definitions from a pluckable file 1`] = `
{
  "definitions": [
    {
      "description": undefined,
      "directives": [],
      "fields": [
        {
          "arguments": [],
          "description": undefined,
          "directives": [],
          "kind": "FieldDefinition",
          "loc": {
            "end": 28,
            "start": 15,
          },
          "name": {
            "kind": "Name",
            "loc": {
              "end": 20,
              "start": 15,
            },
            "value": "hello",
          },
          "type": {
            "kind": "NamedType",
            "loc": {
              "end": 28,
              "start": 22,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 28,
                "start": 22,
              },
              "value": "String",
            },
          },
        },
      ],
      "interfaces": [],
      "kind": "ObjectTypeDefinition",
      "loc": {
        "end": 30,
        "start": 0,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 10,
          "start": 5,
        },
        "value": "Query",
      },
    },
  ],
  "kind": "Document",
  "loc": {
    "end": 30,
    "start": 0,
  },
}
`;

exports[`GitLoader load async should work when loading glob paths that start with './' 1`] = `
{
  "definitions": [
    {
      "description": undefined,
      "directives": [],
      "fields": [
        {
          "arguments": [],
          "description": undefined,
          "directives": [],
          "kind": "FieldDefinition",
          "loc": {
            "end": 28,
            "start": 15,
          },
          "name": {
            "kind": "Name",
            "loc": {
              "end": 20,
              "start": 15,
            },
            "value": "hello",
          },
          "type": {
            "kind": "NamedType",
            "loc": {
              "end": 28,
              "start": 22,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 28,
                "start": 22,
              },
              "value": "String",
            },
          },
        },
      ],
      "interfaces": [],
      "kind": "ObjectTypeDefinition",
      "loc": {
        "end": 30,
        "start": 0,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 10,
          "start": 5,
        },
        "value": "Query",
      },
    },
  ],
  "kind": "Document",
  "loc": {
    "end": 31,
    "start": 0,
  },
}
`;

exports[`GitLoader load sync should load type definitions from a pluckable file 1`] = `
{
  "definitions": [
    {
      "description": undefined,
      "directives": [],
      "fields": [
        {
          "arguments": [],
          "description": undefined,
          "directives": [],
          "kind": "FieldDefinition",
          "loc": {
            "end": 28,
            "start": 15,
          },
          "name": {
            "kind": "Name",
            "loc": {
              "end": 20,
              "start": 15,
            },
            "value": "hello",
          },
          "type": {
            "kind": "NamedType",
            "loc": {
              "end": 28,
              "start": 22,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 28,
                "start": 22,
              },
              "value": "String",
            },
          },
        },
      ],
      "interfaces": [],
      "kind": "ObjectTypeDefinition",
      "loc": {
        "end": 30,
        "start": 0,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 10,
          "start": 5,
        },
        "value": "Query",
      },
    },
  ],
  "kind": "Document",
  "loc": {
    "end": 30,
    "start": 0,
  },
}
`;

exports[`GitLoader load sync should work when loading glob paths that start with './' 1`] = `
{
  "definitions": [
    {
      "description": undefined,
      "directives": [],
      "fields": [
        {
          "arguments": [],
          "description": undefined,
          "directives": [],
          "kind": "FieldDefinition",
          "loc": {
            "end": 28,
            "start": 15,
          },
          "name": {
            "kind": "Name",
            "loc": {
              "end": 20,
              "start": 15,
            },
            "value": "hello",
          },
          "type": {
            "kind": "NamedType",
            "loc": {
              "end": 28,
              "start": 22,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 28,
                "start": 22,
              },
              "value": "String",
            },
          },
        },
      ],
      "interfaces": [],
      "kind": "ObjectTypeDefinition",
      "loc": {
        "end": 30,
        "start": 0,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 10,
          "start": 5,
        },
        "value": "Query",
      },
    },
  ],
  "kind": "Document",
  "loc": {
    "end": 31,
    "start": 0,
  },
}
`;
